lib/monaco-adapter.js

```diff
@@ -1,301 +1,498 @@
+    var otherCursor = this.otherCursors.find(function (cursor) {
+      return cursor.clientID === clientID;
+    });
+
+    /** Initialize empty array, if client does not exist */
+    if (!otherCursor) {
+      otherCursor = {
+        clientID: clientID,
+        decoration: []
+      };
+      this.otherCursors.push(otherCursor);
+    }
 
-      var insert_op = new firepad.TextOperation()
-          .retain(start)
-          .insert(text)
-          .retain(restLength);
+    /** Remove Earlier Decorations, if any, or initialize empty decor */
+    otherCursor.decoration = this.monaco.deltaDecorations(otherCursor.decoration, []);
+    var clazz = "other-client-selection-" + color.replace('#', '');
+    var css;
 
-      var inverse_op = new firepad.TextOperation()
-          .retain(start)
-          .delete(text)
-          .retain(restLength);
+    if (position === selectionEnd) {
+      /** Show only cursor */
+      clazz = clazz.replace('selection', 'cursor');
 
-      return [insert_op, inverse_op];
+      /** Generate Style rules and add them to document */
+      css = getCSS(clazz, 'transparent', color);
+      addStyleRule.call(this, clazz, css);
+    } else {
+      /** Generate Style rules and add them to document */
+      css = getCSS(clazz, color, color);
+      addStyleRule.call(this, clazz, css);
     }
-  };
 
-  MonacoAdapter.prototype.applyOperationToMonaco = function (operation) {
-    var from, index, op, to, _i, _len, _ref, id, ops, range;
-    index = 0;
-    _ref = operation.ops;
-    for (_i = 0, _len = _ref.length; _i < _len; _i++) {
-      op = _ref[_i];
-      if (op.isRetain()) {
-        index += op.chars;
-      } else if (op.isInsert()) {
-        var pos = this.monacoModel.getPositionAt(index);
-        range = new monaco.Range(
-          pos.lineNumber,
-          pos.column,
-          pos.lineNumber,
-          pos.column
-        );
-        id = { major: 1, minor: 1 };
-        ops = {
-          identifier: id, range: range, text: op.text,
-          forceMoveMarkers: true
-        };
-        this.monaco.executeEdits("my-source", [ops]);
-        index += op.text.length;
-      } else if (op.isDelete()) {
-        from = this.monacoModel.getPositionAt(index);
-        to = this.monacoModel.getPositionAt(index + op.chars);
+    /** Get co-ordinate position in Editor */
+    var start = this.monacoModel.getPositionAt(position);
+    var end = this.monacoModel.getPositionAt(selectionEnd);
 
-        range = new monaco.Range(from.lineNumber, from.column, to.lineNumber, to.column);
-        id = { major: 1, minor: 2 };
-        ops = {
-          identifier: id, range: range, text: "",
-          forceMoveMarkers: true
-        };
-        this.monaco.executeEdits("my-source", [ops]);
-      }
+    /** Selection is inversed */
+    if (position > selectionEnd) {
+      var _ref = [end, start];
+      start = _ref[0];
+      end = _ref[1];
     }
-    this.grabDocumentState();
```

Comment:
Do we want to call this at the end of `applyOperation`?

Reply:
No, related state transition has already been taken care of